### PR TITLE
state: move methods to funcs

### PR DESCRIPTION
**Pull request by ebuchman**
State is no longer mutable.

Move all methods mutating methods to be functions that return a new State.

Introduce `BlockExecutor` to handle `ApplyBlock`

### CODE REVIEW

File: evidence/pool.go [hunk 1]
```diff
@@ -9,22 +9,24 @@ import (
...
+	chainID         string
+	lastBlockHeight int64
+	params          types.EvidenceParams
+
 	// never close
 	evidenceChan chan types.Evidence
 }
 
-func NewEvidencePool(params types.EvidenceParams, evidenceStore *EvidenceStore, state types.State) *EvidencePool {
+func NewEvidencePool(params types.EvidenceParams, evidenceStore *EvidenceStore, state *types.State) *EvidencePool {
```
Comment (jaekwon) [inline evidence/pool.go]: Why are we passing a pointer and copying?  Might as well just not receive a pointer...

File: state/state.go [hunk 2]
```diff
@@ -44,9 +41,7 @@ func calcABCIResponsesKey(height int64) []byte {
 // but the fields should only be changed by calling state.SetBlockAndValidators.
 // NOTE: not goroutine-safe.
 type State struct {
-	// mtx for writing to db
-	mtx sync.Mutex
-	db  dbm.DB
+	db dbm.DB
```
Comment (jaekwon) [inline state/state.go]: Can we remove this?

File: state/state.go [hunk 3]
```diff
@@ -82,6 +77,10 @@ type State struct {
 	logger log.Logger
```
Comment (jaekwon) [inline state/state.go]: we should probably remove this too.

File: state/state.go [hunk 4]
```diff
@@ -44,9 +41,7 @@ func calcABCIResponsesKey(height int64) []byte {
 // but the fields should only be changed by calling state.SetBlockAndValidators.
 // NOTE: not goroutine-safe.
```
Comment (jaekwon) [inline state/state.go]: If we're going to have a convention of allowing non-pointer State, (which is nice), then we should also change our State methods to take non-pointer receivers.  The downside is that we can't memoize anything on State anymore, but it looks like we're not doing that anyways.

It's also potentially slower, but I don't think it's enough to worry about, State is overall less than 200 bytes...

File: evidence/pool.go (hunk 1 above)
Comment (melekes) [reply evidence/pool.go]: why are we making a copy? and why not `state.Copy()`?

File: state/execution_test.go [hunk 5]
```diff
@@ -112,15 +55,14 @@ func TestBeginBlockAbsentValidators(t *testing.T) {
 	defer proxyApp.Stop()
 
 	state := state()
-	state.SetLogger(log.TestingLogger())
 
 	// there were 2 validators
-	val1PrivKey := crypto.GenPrivKeyEd25519()
+	/*val1PrivKey := crypto.GenPrivKeyEd25519()
```
Comment (melekes) [inline state/execution_test.go]: m?

File: consensus/replay_test.go [hunk 6]
```diff
@@ -449,31 +453,26 @@ func buildTMStateFromChain(config *cfg.Config, state *sm.State, chain []*types.B
...
 		for _, block := range chain[:len(chain)-1] {
-			applyBlock(state, block, proxyApp)
+			state = applyBlock(stateDB, state, block, proxyApp)
 		}
 
 		// apply the final block to a state copy so we can
 		// get the right next appHash but keep the state back
-		stateCopy := state.Copy()
-		applyBlock(stateCopy, chain[len(chain)-1], proxyApp)
-		latestAppHash = stateCopy.AppHash
```
[the remark this answers is not in the document]
Comment (ebuchman) [reply consensus/replay_test.go]: no we propagate the state instead, which contains the app hash

File: evidence/pool.go [hunk 7]
```diff
@@ -3,29 +3,32 @@ package evidence
...
-func NewEvidencePool(params types.EvidenceParams, evidenceStore *EvidenceStore, state types.State) *EvidencePool {
+func NewEvidencePool(params types.EvidenceParams, evidenceStore *EvidenceStore, state sm.State) *EvidencePool {
 	evpool := &EvidencePool{
+		state:         state,
 		params:        params,
 		logger:        log.NewNopLogger(),
 		evidenceStore: evidenceStore,
-		state:         state,
-		evidenceChan:  make(chan types.Evidence),
+		// state:         *state,
```
[the remark this answers is not in the document]
Comment (ebuchman) [reply evidence/pool.go]: yes thanks

File: state/execution_test.go (hunk 5 above)
Comment (ebuchman) [reply state/execution_test.go]: right will look more closely. it turned out we weren't using this, but maybe we should have been

**Pull request merged**